client/block.c

```diff
@@ -1941,16 +1941,36 @@ void xdag_block_finish()
 	pthread_mutex_lock(&block_mutex);
 }
 
-int xdag_get_block_info(xdag_hash_t hash, void *data, int (*callback)(void*, int, xdag_hash_t, xdag_amount_t, xtime_t, const char *))
+int xdag_get_block_info(xdag_hash_t hash, void *info, int (*info_callback)(void*, int, xdag_hash_t, xdag_amount_t, xtime_t, const char *),
+						void *links, int (*links_callback)(void*, const char *, xdag_hash_t, xdag_amount_t))
 {
 	pthread_mutex_lock(&block_mutex);
 	struct block_internal *bi = block_by_hash(hash);
 	pthread_mutex_unlock(&block_mutex);
 
-	if(callback && bi) {
-		return callback(data, bi->flags & ~BI_OURS,  bi->hash, bi->amount, bi->time, get_remark(bi));
+	if(info_callback && bi) {
+		info_callback(info, bi->flags & ~BI_OURS,  bi->hash, bi->amount, bi->time, get_remark(bi));
 	}
 
+	if(links_callback && bi) {
+		int flags;
+		struct block_internal *ref;
+		pthread_mutex_lock(&block_mutex);
```

Comment:
Why do you need mutex here? Do you think fields 'ref' and 'flags' can be modified for received block? If so, why didn't you use mutex for lines 1952, 1968, 1971?

Reply:
i did it in the function the prints block info
you need to take flags and ref together to be safe so the murex is needed.
without mutex you can see flag BI_REF and old ref (that is oref) and you can dereference ref as block_internal but it may be a orphan_block! and you may get a seg fault.
so if BI_REF is set and ref is not 0 is safe (but you need to take them together atomically... so you need the mutex)
(a way to avoid mutex was to make flags an atomic variable but i didn't want to add so much overhead...)
1952 is safe because there is no dereference except bi-> that is safe.
1968 is safe for the same reason.
1971 is not safe... !

Reply:
fixed it for function printf_blocks in #431

Reply:
I will port your fix to this PR. @sgaragagghu